File: superstitioMain/java/superstitio/cards/general/SkillCard/hangUpCard/OneMoreHour.java
package superstitio.cards.general.SkillCard.hangUpCard;

import com.megacrit.cardcrawl.characters.AbstractPlayer;
import com.megacrit.cardcrawl.monsters.AbstractMonster;
import superstitio.DataManager;
import superstitio.cards.general.GeneralCard;
import superstitio.delayHpLose.RemoveDelayHpLoseBlock;
import superstitioapi.hangUpCard.CardOrb;
import superstitioapi.hangUpCard.Card_TriggerHangCardManually;

import static superstitioapi.InBattleDataManager.getHangUpCardOrbGroup;
import static superstitioapi.actions.AutoDoneInstantAction.addToBotAbstract;

public class OneMoreHour extends GeneralCard implements Card_TriggerHangCardManually {
    public static final String ID = DataManager.MakeTextID(OneMoreHour.class);

    public static final CardType CARD_TYPE = CardType.SKILL;

    public static final CardRarity CARD_RARITY = CardRarity.COMMON;

    public static final CardTarget CARD_TARGET = CardTarget.SELF;

    private static final int COST = 1;
    private static final int BLOCK = 7;
    private static final int UPGRADE_BLOCK = 3;
    private static final int MAGIC = 2;
    private static final int UPGRADE_MAGIC = 1;

    public OneMoreHour() {
        super(ID, CARD_TYPE, COST, CARD_RARITY, CARD_TARGET);
        this.setupBlock(BLOCK, UPGRADE_BLOCK, new RemoveDelayHpLoseBlock());
        this.setupMagicNumber(MAGIC, UPGRADE_MAGIC);
    }

    @Override
    public void use(AbstractPlayer player, AbstractMonster monster) {
        addToBot_gainBlock();
        for (int i = 0; i < this.magicNumber; i++) {
            addToBotAbstract(() -> getHangUpCardOrbGroup()
                    .ifPresent(cardGroup -> cardGroup.cards
                            .forEach(orb -> orb.OrbCounter *= this.magicNumber)));
        }
    }

    @Override
    public void upgradeAuto() {
    }

    @Override
    public boolean forceFilterCardOrbToHoveredMode(CardOrb orb) {
        orb.targetType = CardOrb.HangOnTarget.None;
        orb.actionType = CardOrb.HangEffectType.Good;
        return true;
    }

    @Override
    public int forceChangeOrbCounterShown(CardOrb orb) {
        return orb.OrbCounter * this.magicNumber;
    }
}

File: superstitioMain/java/superstitio/cards/general/SkillCard/hangUpCard/OneMoreHour.kt
package superstitio.cards.general.SkillCard.hangUpCard

import com.megacrit.cardcrawl.characters.AbstractPlayer
import com.megacrit.cardcrawl.monsters.AbstractMonster
import superstitio.DataManager
import superstitio.cards.general.GeneralCard
import superstitio.delayHpLose.RemoveDelayHpLoseBlock
import superstitioapi.hangUpCard.CardOrb
import superstitioapi.hangUpCard.Card_TriggerHangCardManually
import superstitioapi.hangUpCard.HangUpCardGroup

class OneMoreHour : GeneralCard(ID, CARD_TYPE, COST, CARD_RARITY, CARD_TARGET), Card_TriggerHangCardManually
{
    init
    {
        this.setupBlock(BLOCK, UPGRADE_BLOCK, RemoveDelayHpLoseBlock())
        this.setupMagicNumber(MAGIC, UPGRADE_MAGIC)
    }

    override fun use(player: AbstractPlayer?, monster: AbstractMonster?)
    {
        addToBot_gainBlock()
        //        for (int i = 0; i < this.magicNumber; i++) {
        HangUpCardGroup.forEachHangUpCard { it.orbCounter += this.magicNumber }
            .addToBotAsAbstractAction()
        //        }
    }

    override fun upgradeAuto()
    {
    }

    override fun forceFilterCardOrbToHoveredMode(orb: CardOrb): Boolean
    {
        orb.targetType = CardOrb.HangOnTarget.None
        orb.actionType = CardOrb.HangEffectType.Good
        return true
    }

    override fun forceChangeOrbCounterShown(orb: CardOrb): Int
    {
        return orb.orbCounter.toInt { it + this.magicNumber }
    }

    companion object
    {
        val ID: String = DataManager.MakeTextID(OneMoreHour::class.java)

        val CARD_TYPE: CardType = CardType.SKILL

        val CARD_RARITY: CardRarity = CardRarity.COMMON

        val CARD_TARGET: CardTarget = CardTarget.SELF

        private const val COST = 1
        private const val BLOCK = 7
        private const val UPGRADE_BLOCK = 3
        private const val MAGIC = 3
        private const val UPGRADE_MAGIC = 1
    }
}
